### DISCUSSION TITLE
GrandOrgue, Musescore and MIDI

### DISCUSSION THREAD

**Discussion by newVideophile** [GrandOrgue/grandorgue#1225] [Help]
I'm not sure if there are people knowing both GrandOrgue and Musescore here, but I have the following problem : I have written a composition in Musescore for organ. When I export it as a midi file, and then import it in GrandOrgue, everything will be played on the great. 

I didn't find any menu to set midi parameters, neither for Musescore export nor for GrandOrgue import. 

Any help will be appreciated!

**Comment by oleg68** [comment]
Usually different manuals are assigned to different MIDI channels in GrandOrgue. So for playing on several manuals the MIDI file has to contain midi events distributed to several midi channels.

Musescore is mostly a score-editing software, not a midi sequencing software. There are some specialised midi editors. The most famous is Rosegarden. It allows to edit the mid file exported from Musescore and to distribute the events acress different channels.

**Reply by newVideophile** [reply]
Thank you oleg. I found a workaround in Musescore: I select 3 organs, then I delete 2 staves on each organ, in order to get 3 staves seen as 3 different instruments - and thus 3 different midi channels - by Musescore. 
Now, when played on GrandOrgue, my file will use the solo, the swell and the great. This is better, but not what I want. I would like to use the swell, the great and the pedal. How to tell that to GrandOrgue?

**Reply by oleg68** [reply]
1. Right ckick on the swell, great, pedal and look at the MIDI channel assigned to each manual.
2. Make sure that your midi file contain that channel numbers.

**Reply by newVideophile** [reply]
I managed to set the midi channels using the mixer. Musescore is very powerfull but somethimes not intuitive: I would have expected a configuration panel while exporting to midi. 
Anyway, it works, many thanks for your help !

**Comment by midimusic** [comment]
You can use eplayOrgan to play your midi files, download the new version from my website. You will find there are many options available either for recording. playing your files direct or by routing them from eplayOrgan to your favourite GO organ. 
csw900

**Reply by newVideophile** [reply]
Thank you for the suggestion. Are you the author of eplayOrgan? It looks impressive!

I'm not sure how eplayOrgan could help here, though. All a need is to be able to write a score in Musescore and then to be able to play it with GrandOrgue. 

So far, I have managed to do it with only a few limitations:
1) tempo changes cause desynchronization between channels in GrandOrgue (not in Musescore)
2) the transfer from Musescore to GrandOrgue requires 15 clicks (save the file in Musescore, export to midi, confirm overwriting the file, transfer to the GrandOrgue system using Filezilla, confirm overwriting again, load the midi file in GrandOrgue, and finally press play)

Could eplayOrgan help solving these problems?

**Reply by midimusic** [reply]
Yes I am the author of eplayOrgan. You can email me direct if you have problems with it.
1. How do you set the stops on your compositions?
2. You are probably not doing your tempo changes correctly - you have to change the tempo on ALL channels at the same moment. Organ midi files very rarely contain midi tempo changes - it is usually the organist that plays at a different speed without having any formal tempo change. Midi tempo changes are not a good idea and should be avoided apart from setting the tempo at the start of the midi file.
3. eplayOrgan can automatically convert your midi files to GO format. I assume you are using Filezilla for this. It can also play your files in a more flexible manner than GO. "GO is not a midi file player" - Quote from Lars Palo.
csw900

**Reply by newVideophile** [reply]
To answer your questions:
1. I don't set the stops in Musescore because I don't know how to do that. To get a preview of my composition, I let GrandOrgue play the MIDI file while I manually recall my registrations at the appropriate moments. Being the assistant of a machine is quite funny!
2. Musescore applies a tempo change to all staves/channels at the same time. When playing in Musescore it works. When exporting to midi, then playing the midi file in Musescore, it works. But the same midi file doesn't work in GO. I tried to apply the same tempo change to each of the 3 staves, but it didn't work. My problem may be related to the fact that I blend three organs in Musescore, because if I use the standard provided organ, although it has 3 staves as expected, they will all play on the same midi channel. Regarding your comment about tempo changes: I know that the organist interprets the composition with many effects not written on the score, but I would like the midi file to provide me with a semi-realistic preview, including tempo changes between sections and slowing down at the end. A composition with the same tempo from beginning to end is not music. A Mario kart sound track at best ;-)
3. I will try that, thank you for the suggestion

**Reply by midimusic** [reply]
In my experience you will get the best results by using pistons to insert the stops while the tune is playing. You can also insert them with a midi file editor (once you have learned how).

It does not surprise me that your tempo changes do not work properly. Rather than using Musescore to "blend" the three organs I think you ought to use a midi file editor on the Musescore's output to change its single channel to three distinct channels.

It would also be a better idea to use the midi editor to insert your tempo changes. I have no idea how GO will react to tempo changes - as I said above organ midi files do not normally contain them.
csw900

**Reply by newVideophile** [reply]
"In my experience you will get the best results by using pistons to insert the stops while the tune is playing" --> yes, this is exactly what I am doing

"It does not surprise me that your tempo changes do not work properly" --> I have done further tests. The same midi file will play correctly in Musescore and eplayOrgan, but not in GO. A bug in GO?

"organ midi files do not normally contain [tempo changes]" --> why not?

**Reply by midimusic** [reply]
Organ midi files never need midi tempo changes because:
1 The organist has no facilities to insert them while he is playing, even if he wanted to.
2. They are never required because the organist can change tempo simply by changing the speed he is playing. The midi file records these speed changes without needing any specific (tempo change) instruction to do so. You could simulate this in Musescore, but not as easily as the organist can do it without even thinking.
3. No it is not a bug in GO for the reasons above.
csw900

**Reply by newVideophile** [reply]
We are talking about different situations. You describe the recording of a midi file using an organ equipped with a midi output. My usage is different. I have composed a piece in Musescore (on my PC only, whithout any midi input device), and I needed a reasonably realistic preview in order to validate it. My goal is to print the score (done), learn it (in progress) and play it live (planned). 

Regarding GO, I do think that there is a bug. Here is a test score created in Musescore:
![Test_tempo](https://user-images.githubusercontent.com/95248680/198015448-e9d5fb98-fee7-44a3-8a03-532825ab6baa.jpg)

When I convert it to midi and play it in Musescore and eplayOrgan, I get a correct result, whereas GO will play it incorrectly: all keyboards should play the same note at the same time, and it is not the case. Here is a screenshot of GO while playing this midi file that shows the problem:

![GO](https://user-images.githubusercontent.com/95248680/198016603-12c2068a-d840-497d-8286-dc424e19b7ca.jpg)

Here, both the great and the pedal are playing a G, but the swell is playing an E... Surprisingly, the desynchronization occurs from the beginning, not only after the tempo change.

**Reply by oleg68** [reply]
> Here, both the great and the pedal are playing a G, but the swell is playing an E... Surprisingly, the desynchronization occurs from the beginning, not only after the tempo change.

Could you attach here the midi file reproducing this problem?

**Reply by newVideophile** [reply]
Yes, of course: [Test_tempo_midi.zip](https://github.com/GrandOrgue/grandorgue/files/9869346/Test_tempo_midi.zip)

**Reply by midimusic** [reply]
I have had a quick look at your midi file and immediately see that two of the channels have "Reset all controllers" as the first event whereas the other channel has an "undefined" event in the same position. This is probably what is confusing GO.
Also:
1. I see "breath control" event on each note - what is this doing in an organ midi file?
2. There are three other unusual events RPN "Pitch Bend", + two "Effects Depth" events. These should not be in an organ midi file.
Both 1. and 2. will/should be ignored by an organ.

Also I point out that organs are unlikely to recognise key or time signatures. They will/should be ignored.

I will be interested in Oleg's comments.
csw900

**Comment by midimusic** [comment]
I have been exerimenting with the latest version of musescore and have come to the conclusion that instead of using musescore to make a midi recording you should setup musescore to output midi and use a proper organ to record your midi.
Musescore has an I/O option to output midi to either a real or virtual midi cable. Use this output to connect to an organ, play your score and record it on the organ.
There is a problem that I cannot get musescore to use this midi output. There are many videos on youtube suggesting that this output may not work. However these videos suggest using the JACK output may work OK.
I suggest you experiment yourself as I do not want to install JACK on my windows system.
csw900
Later...
I have now found out how to properly play Musescore files on an organ!!
1. Load your musescore file into musescore.
2. Use the musescore file menu to export a MIDI file.
3. Edit the channels in the midi file using a midi editor so they match the organ you want to play. I used eplayOrgan and a very old version of Cakewalk PA for editing and playing. Remove any presets (set them to -1).
4. Set up the midi editor to play the file to a midi output (Use a virtual midi cable such as Loopbe). The midi editor will correctly play any tempo changes so they will be played and record correctly.
5. Connect your organ's midi input to the other end of the virtual midi cable.
6. Play the file on the midi editor - its output will play the organ. You can set appropriate stops and record the result on the organ's midi recorder.

**Reply by newVideophile** [reply]
This works probably, but it would adds more steps to my already too complex workflow, so I'm not really motivated to try that. I prefer to live with the limitation of a fixed tempo. My need is to get a preview, not an artistic masterpiece. But thank you for the help anyway!

**Comment by oleg68** [comment]
Ffter decoding this midi file, I found that it contains three tracks.

- Tempo: 100
- First track:
  - D
  - E
  - F#
  - G
  - Tempo: 50
  - D
  - E
  - F#
  - G
- Second track:
  - D
  - E
  - F#
  - G
  - Tempo: 50
  - D
  - E
  - F#
  - G
- Third track:
  - D
  - E
  - F#
  - G
  - Tempo: 50
  - D
  - E
  - F#
  - G

GrandOrgue assumes that every midi tempo event (51) influences to all following midi events regardless the track. So the first 4 notes from the first track are played at 100 bpm and all notes from another tracks are played at 50 bpm.

Seems another midi players essume that the midi tempo events (51) influence to the current track only, so they play the first 4 notes from each track at 100 bpm and the second 4 notes at 50 bpm.

@midimusic What is the right scope of the tempo event (51): the current track or the whole file?

**Reply by newVideophile** [reply]
oleg, I don't understand that. If a midi tempo event applies to all tracks, then why doesn't the first "tempo = 100" event cause all tracks to play the first 4 notes at 100 bpm?

Moreover, how can the "tempo=50" events apply to events that occured _before_?

**Reply by oleg68** [reply]
> Moreover, how can the "tempo=50" events apply to events that occured before?

There are two different `before`:
- before at the time
- before in the file

Because tracks are recorded in the file sequentially: the whole first track, then the whole second track, then the whole third track, so GO applies 50 to all events after in the file, so to the whole second and third tracks

**Reply by newVideophile** [reply]
Ah ok! Hence your previous question regarding the scope of a midi tempo event.

Well, this discussion [https://www.midi.org/forum/14749-tempo-map-and-midi-files-of-format-1](url) explains that, according to the midi standard, all tempo events should be placed in track 1, although many midi files do not respect this requirement. I deduce that a tempo event is global, not limited to a track. So the way GrandOrgue interprets midi should be modified on this aspect.

**Reply by oleg68** [reply]
But from http://midi.teragonaudio.com/tech/midifile/tempo.htm

> In a format 0 file, the tempo changes are scattered throughout the one MTrk. In format 1, the very first MTrk should consist of only the tempo (and time signature) events so that it could be read by some device capable of generating a "tempo map". It is best not to place MIDI events in this MTrk. In format 2, each MTrk should begin with at least one initial tempo (and time signature) event.

So I deduce that a tempo event should affect only one track. GO should remember the initial tempo from the midi file header and should reset to it at the beginning of each track.

**Reply by midimusic** [reply]
Let us get some sense into this discussion.

Midi files are always played in the order of the TIME the event occured. Thus a file with three tracks has to be sorted into time order before it can be played. Maybe GO is not doing this - if so it can never play midi files with more than one track.

Thus midi tempo events will/should be processed according to WHEN they occured which has nothing to do with their position in the file.

Normally an organ midi file contains only one tempo event at the beginning which affects ALL tracks/channels at the time it occurs. Any subsequent tempo events will affect all tracks/channels at the time they occur. (If there is no tempo event then the midi default of 120 bps is used)
csw900

**Reply by oleg68** [reply]
The problem is the TIME of events is calculated from the tempo. So GO cann't sort the events before it uses the correct tempo.

I've just changed GrandOrgue to inherit tempo changes from the first track. Now the file provided by @newVideophile is played correctly.

**Reply by midimusic** [reply]
Time is relative. The file should be sorted before any other processing takes place. Absolute times are then set according to the tempo.
All tempo events apply to all tracks whenever and wherever the occur.
How many years has it taken to discover this bug? This may be a record.
csw900

**Reply by newVideophile** [reply]
> Time is relative

As Albert said! :-D

**Reply by newVideophile** [reply]
> I've just changed GrandOrgue to inherit tempo changes from the first track. Now the file provided by @newVideophile is played correctly.

Many thanks oleg!

**Reply by oleg68** [reply]
You can try [the intermediate GrandOrgue build](https://github.com/oleg68/GrandOrgue-official/actions/runs/3361331688) with corrected processing of midi tempo changes.

**Reply by newVideophile** [reply]
oleg, I'm not familiar with Github, how to install your build? I did'nt find a Windows install exe or zip file from your link.

**Reply by rousseldenis** [reply]
On the bottom of the page at 'Artifacts' title, you'll find files to download for your architecture.

**Reply by kerkovits** [reply]
BTW, version 3.9.0 should include this fix.

**Reply by rousseldenis** [reply]
Indeed

**Reply by newVideophile** [reply]
> You can try [the intermediate GrandOrgue build](https://github.com/oleg68/GrandOrgue-official/actions/runs/3361331688) with corrected processing of midi tempo changes.

Ok, so I managed to install the intermediate build (for some reason the download link was not working initially). 

It works on the simple file I provided, but still doesn't work on more complex pieces. Oleg, I can send a new example midi file to you, but I would prefer to do it privately, is there a way to do that with Github?

**Reply by oleg68** [reply]
> I can send a new example midi file to you, but I would prefer to do it privately, is there a way to do that with Github?

No, there is not. Github does not provide this capability.

But you can send me your midi file with e-mail. Find my e-mail address at the line 26 of https://github.com/GrandOrgue/grandorgue/blob/master/AUTHORS

**Reply by newVideophile** [reply]
Done!

**Comment by midimusic** [comment]
Here is a nice one to try: (edit the channels to suit your organ)
[Toccata_and_Fugue_BWV_565_in_D_Minor.zip] (https://github.com/GrandOrgue/grandorgue/files/9949697/Toccata_and_Fugue_BWV_565_in_D_Minor.zip)
csw900

**Comment by midimusic** [comment]
@oleg68 
Every midi tempo meta event (ff 51 03 tttttt) is associated with a time and affects every track and channel from the time it is received. Implementing this is quite tricky as the time of every subsequent event has to be appropriately updated.
@newVideophile  and anyone else interested.
I have just issued a new version 4.2.0.3 of eplayOrgan64 which can play any midi file exported by MuseScore at the correct tempo throughout. It shows time and key signatures and a time/tempo map of the file. It also actively displays the current tempo whilst playing the file.
You can test it out with the Toccata midi file above or any other MuseScore midi file.
Please let me know if any other features would be useful to an organ music composer and I will see if I can implement them.
csw900

**Reply by ministerinmarin** [reply]
Hi @midimusic I created comments on this issue using a different Muse Score encode. Please see below.

**Comment by ministerinmarin** [comment]
Currently investigating this with an organ adaptation (originally trumpets and keyboard) of a Christmas carol arranged by Don Hustad, first page of original and adapted attached below.
![image](https://user-images.githubusercontent.com/104122030/211186756-9ff382c0-1796-4ec7-8d7f-af5f20dba13c.png)
![image](https://user-images.githubusercontent.com/104122030/211186774-84212b09-a13b-4fa8-9817-fcff4a6874ea.png)

Every VPO will interpret midis differently, depending on the number of manual(s) and pedal available to the VPO, as well as how you arranged the staffs in MuseScore, and sometimes depending on the notes themselves (It is hard for me to explain that further), specifically with keyboard midi files using the grand staff of treble and bass.

I attempted that to work on the organs of Pitea School of Music, Pitea positive, and Barton 3-7 theater organ. Here's the implementation using Pitea School of Music Organ:
![image](https://user-images.githubusercontent.com/104122030/211186980-147153a7-ab28-40ed-8620-28ff15d80b82.png)
![image](https://user-images.githubusercontent.com/104122030/211186991-4ea4d28c-7fe4-4569-9224-b4e1f86b0317.png)
[Audio snippet (with rather bad choice of stops, as I am new to VPOs but have been a pipe organ appreciator for many years)](https://cdn.discordapp.com/attachments/859441862535151637/1061561345927037008/2023-01-08-16-20-04.547.wav)

**Reply by ministerinmarin** [reply]
![image](https://user-images.githubusercontent.com/104122030/211187201-d832f902-2e0c-4e8c-b061-f317ff488b56.png)
This is how it would look like when reimporting the midi to MuseScore
The staffs will expand from four to seven

**Reply by ministerinmarin** [reply]
[Current MuseScore reworking to organ from trumpet and keyboard arrangement](https://cdn.discordapp.com/attachments/859441862535151637/1061598829373292544/o_holy_night_organ.mscz)
[Current export as MIDI render](https://cdn.discordapp.com/attachments/859441862535151637/1061598932980994098/o_holy_night_organ_2.mid)
[Earlier MIDI render](https://cdn.discordapp.com/attachments/859441862535151637/1061598908851179580/o_holy_night_organ.mid)

[How it is supposed to sound, MP3 export from MuseScore using Burea Church Choir Organ sf2](https://cdn.discordapp.com/attachments/859441862535151637/1061598968791969822/O_Holy_Night_Arr._Hustad.mp3)
[Original score for SATB choir, keyboard, and three trumpets](https://cdn.discordapp.com/attachments/859441862535151637/1061600724376633344/OHolyNight.pdf)
[Original working from 2019](https://cdn.discordapp.com/attachments/859441862535151637/1061602745863381032/o_holy_night_ORIGINAL.mscz)

**Reply by ministerinmarin** [reply]
Oh, before I forget, sometimes MuseScore voices 2 to 4 won't work very well in some organs... where it would be challenging for the music editor to implode them into one or two staffs and have it recognized in both GO and eplayOrgan, according to my observation.
![image](https://user-images.githubusercontent.com/104122030/211193104-9beb300f-34dc-4f61-aaa5-35a88aa59e01.png)

**Comment by ministerinmarin** [comment]
> I didn't find any menu to set midi parameters, neither for Musescore export nor for GrandOrgue import.

![image](https://user-images.githubusercontent.com/104122030/211187326-654ff700-f18f-45b4-a9a0-2ea8435b4fe4.png)
Have you tried the piano roll editor? Select a measure and right-click and find the "Piano Roll Editor" in the menu.

**Comment by midimusic** [comment]
@ministerinmarin

You have not said which version of Musescore you are using - they are very different. Neither of them works particularly well with organ midi. The authors do not seem to understand that organs need more than one midi channel.

For playing Musescore midi on an organ the only useful way is to export a midi file and then edit the midi channels to suit your organ.

Regarding the import of midi files into Musescore - that is a Musescore problem. Currently I would be surprised if it worked properly with organ music.

Grand Orgue midi player is designed to work only with midi files made by Grand Orgue. Other midi files can be played but they must be made to look like GO files in order to play properly. (eplayOrgan works much better with midi files than GO).

csw900

**Reply by ministerinmarin** [reply]
I am using MuseScore 3.6.2, which is the last version before the 4.0 release (I also have it, but that is problematic to me currently). The original was imported from MuseScore 2.3.2 where I encoded the file back in 2019. The 2023 MuseScore reworking can be downloaded [from this link](https://cdn.discordapp.com/attachments/859441862535151637/1061598829373292544/o_holy_night_organ.mscz).

But you are correct, specific implementation of organ MIDI, for instance for playing back on GO *and* on instruments that support MIDI but don't have GO, will require a MIDI rewiring outside MuseScore. There's no interoperability here besides the usual General MIDI support.

**Reply by ministerinmarin** [reply]
![image](https://user-images.githubusercontent.com/104122030/211197669-f7f6cf8d-0ac9-4486-95ff-96d6d22b6102.png)
> Regarding the import of midi files into Musescore - that is a Musescore problem. Currently I would be surprised if it worked properly with organ music.

Here's a GO MIDI import into MuseScore 4. Like other keyboard instruments supporting MIDI recording, there is no note consistency in relation to the measures and tempo. And every top note has this text ``+ 5 [54]``.

**Comment by midimusic** [comment]
@ministerinmarin 

There is another problem you may not have discovered yet. Musescore defaults to using cc2's for expression control in midi files. I do not know of any organ that  uses this - it needs to be changed to cc11 before exporting midi files. Musescore 3 allows this change to be done but Musescore 4 does not!!

csw900
